File: 1552-magnetic-force-between-two-balls/1552-magnetic-force-between-two-balls.py
class Solution:
    def maxDistance(self, pos: List[int], m: int) -> int:
        
        def allocate(mid):
            ball = 1
            last = pos[0]
            for i in range(1, len(pos)):
                if pos[i]-last>=mid:
                    ball+=1
                    last = pos[i]
            return ball
        
        
        pos.sort()
        lo, hi, res = 1, pos[-1], 1
        
        
        while lo<=hi:
            mid = (lo+hi)//2
            
            ball = allocate(mid)
            if ball>=m:
                res = mid
                lo = mid + 1
            else:
                hi = mid -1
        
        return res

File: 1552-magnetic-force-between-two-balls/1552-magnetic-force-between-two-balls.java
class Solution {
    public int maxDistance(int[] pos, int m) {
        Arrays.sort(pos);
        int lo = 1;
        int hi = pos[pos.length-1] - pos[0];
        int ans = -1;
        
        while(lo <= hi){
            int mid = lo + (hi - lo)/2;
            if(valid(pos, m, mid)){
                ans = mid;
                lo = mid + 1;
            }
            else
                hi = mid - 1;
        }
        return ans;
    }
    
    boolean valid(int[] pos, int m, int d){
        int last = pos[0];
        for(int i=1; i<pos.length; ++i){
            if(pos[i]-last >= d){
                --m;
                last = pos[i];
            }
        }
        return m<=1;
    }
}